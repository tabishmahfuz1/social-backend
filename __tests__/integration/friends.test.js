const { gql } = require("apollo-server");
const faker = require("faker");
const { testClient } = require("../util/testClient");
const typeDefs = require("../../src/graphql/typeDefs");
const resolvers = require("../../src/graphql/resolvers/index");
const User = require("../../src/models/User");
const generateToken = require("../../src/util/generateToken");

const ADD_FRIEND = gql`
  mutation addFriend($notifier: String!) {
    addFriend(notifier: $notifier) {
      creator {
        firstName
        lastName
        avatarImage
      }
      notifier {
        firstName
        lastName
        avatarImage
      }
      action
      status
    }
  }
`;

const ACCEPT_FRIEND = gql`
  mutation acceptFriend($creator: String!) {
    acceptFriend(creator: $creator) {
      creator {
        firstName
        lastName
        avatarImage
      }
      notifier {
        firstName
        lastName
        avatarImage
      }
      action
      status
    }
  }
`;

const REJECT_FRIEND = gql`
  mutation rejectFriend($creator: String!) {
    rejectFriend(creator: $creator)
  }
`;

const REMOVE_FRIEND = gql`
  mutation removeFriend($creator: String!) {
    removeFriend(creator: $creator) {
      id
      friends {
        id
        firstName
        lastName
        avatarImage
        username
      }
    }
  }
`;

describe("Friends integration testing", () => {
  test("should create a notification with status: pending if User A sends a friend request to User B", async () => {
    // ARRANGE
    const userA = await new User({
      firstName: "Vincent",
      lastName: "Beahan",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const userB = await new User({
      firstName: "Emil",
      lastName: "Reinger",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const token = generateToken(userA);
    // ACT
    // use the test server to create a query function
    const { mutate } = testClient(
      {
        resolvers,
        typeDefs,
        context: async ({ req }) => ({ req }),
      },
      {
        req: { headers: { authorization: `JWT ${token}` } },
      }
    );
    // send a friend request from userA to userB
    const notification = await mutate({
      mutation: ADD_FRIEND,
      variables: {
        notifier: userB.username,
      },
    });

    // ASSERT
    expect(notification).toMatchObject({
      data: {
        addFriend: {
          status: "pending",
          creator: {
            firstName: userA.firstName,
            lastName: userA.lastName,
          },
          notifier: {
            firstName: userB.firstName,
            lastName: userB.lastName,
          },
        },
      },
    });
    expect(notification).toMatchSnapshot();
  });
  test("User A sends a friend request to User B => if User B accepts add User A to User B's friends array", async () => {
    // ARRANGE
    const userA = await new User({
      firstName: "Vincent",
      lastName: "Beahan",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const userB = await new User({
      firstName: "Emil",
      lastName: "Reinger",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const token = generateToken(userA);
    // ACT
    // use the test server to create a query function
    const { mutate } = testClient(
      {
        resolvers,
        typeDefs,
        context: async ({ req }) => ({ req }),
      },
      {
        req: { headers: { authorization: `JWT ${token}` } },
      }
    );
    // send a friend request from userA to userB
    await mutate({
      mutation: ADD_FRIEND,
      variables: {
        notifier: userB.username,
      },
    });

    const tokenB = generateToken(userB);

    // accept the friend request from userB
    const acceptNotification = await mutate({
      mutation: ACCEPT_FRIEND,
      variables: {
        creator: userA.username,
      },
      ctxArg: { req: { headers: { authorization: `JWT ${tokenB}` } } },
    });
    expect(acceptNotification).toMatchSnapshot();

    // ASSERT
    expect(acceptNotification).toMatchObject({
      data: {
        acceptFriend: {
          status: "accepted",
          creator: {
            firstName: userA.firstName,
            lastName: userA.lastName,
          },
          notifier: {
            firstName: userB.firstName,
            lastName: userB.lastName,
          },
        },
      },
    });
    // assert that user B and user A have their ids in each one's friends array
    const newUserA = await User.findById(userA.id);
    const newUserB = await User.findById(userB.id);

    expect(newUserA.friends[0].toString()).toEqual(newUserB.id);
    expect(newUserB.friends[0].toString()).toEqual(newUserA.id);
  });
  test("User A sends a friend request to User B => if User B rejects then remove the friend notification", async () => {
    // ARRANGE
    const userA = await new User({
      firstName: "Vincent",
      lastName: "Beahan",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const userB = await new User({
      firstName: "Emil",
      lastName: "Reinger",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const token = generateToken(userA);
    // ACT
    // use the test server to create a query function
    const { mutate } = testClient(
      {
        resolvers,
        typeDefs,
        context: async ({ req }) => ({ req }),
      },
      {
        req: { headers: { authorization: `JWT ${token}` } },
      }
    );
    // send a friend request from userA to userB
    await mutate({
      mutation: ADD_FRIEND,
      variables: {
        notifier: userB.username,
      },
    });

    const tokenB = generateToken(userB);

    // reject the friend request from userB
    await mutate({
      mutation: REJECT_FRIEND,
      variables: {
        creator: userA.username,
      },
      ctxArg: { req: { headers: { authorization: `JWT ${tokenB}` } } },
    });

    // assert that user B and user A do NOT have their ids in each one's friends array
    const newUserA = await User.findById(userA.id);
    const newUserB = await User.findById(userB.id);

    expect(newUserA.friends[0]).toBeUndefined();
    expect(newUserB.friends[0]).toBeUndefined();
  });
  test("User A is friends with User B => if User B removes User A from his friendlist then remove each ids from each user's friends array", async () => {
    // ARRANGE
    const userA = await new User({
      firstName: "Vincent",
      lastName: "Beahan",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const userB = await new User({
      firstName: "Emil",
      lastName: "Reinger",
      email: faker.internet.email(),
      password: faker.internet.password(),
      gender: faker.random.arrayElement(["male", "female"]),
    }).add();

    const token = generateToken(userA);
    // ACT
    // use the test server to create a query function
    const { mutate } = testClient(
      {
        resolvers,
        typeDefs,
        context: async ({ req }) => ({ req }),
      },
      {
        req: { headers: { authorization: `JWT ${token}` } },
      }
    );
    // send a friend request from userA to userB
    await mutate(
      {
        mutation: ADD_FRIEND,
        variables: {
          notifier: userB.username,
        },
      },
      {
        req: { headers: { authorization: `JWT ${token}` } },
      }
    );

    const tokenB = generateToken(userB);

    // accept the friend request from userB
    await mutate({
      mutation: ACCEPT_FRIEND,
      variables: {
        creator: userA.username,
      },
      ctxArg: { req: { headers: { authorization: `JWT ${tokenB}` } } },
    });

    // remove the friend ( User A ) from userB
    await mutate({
      mutation: REMOVE_FRIEND,
      variables: {
        creator: userA.username,
      },
      ctxArg: { req: { headers: { authorization: `JWT ${tokenB}` } } },
    });

    // assert that user B and user A do NOT have their ids in each one's friends array
    const newUserA = await User.findById(userA.id);
    const newUserB = await User.findById(userB.id);

    expect(newUserA.friends[0]).toBeUndefined();
    expect(newUserB.friends[0]).toBeUndefined();
  });
});
