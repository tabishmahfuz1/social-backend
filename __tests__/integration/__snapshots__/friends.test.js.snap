// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friends integration testing User A sends a friend request to User B => if User B accepts add User A to User B's friends array 1`] = `
Object {
  "data": Object {
    "acceptFriend": Object {
      "action": "Sent you a friend request",
      "creator": Object {
        "avatarImage": "https://scontent.fsof4-1.fna.fbcdn.net/v/t1.30497-1/c71.0.240.240a/p240x240/84241059_189132118950875_4138507100605120512_n.jpg?_nc_cat=1&_nc_sid=7206a8&_nc_ohc=CD3ImewvDlgAX8f-QvY&_nc_ht=scontent.fsof4-1.fna&oh=782172992e9272da49f35f7f40ef8d63&oe=5EED0201",
        "firstName": "Vincent",
        "lastName": "Beahan",
      },
      "notifier": Object {
        "avatarImage": "https://scontent.fsof4-1.fna.fbcdn.net/v/t1.30497-1/c71.0.240.240a/p240x240/84241059_189132118950875_4138507100605120512_n.jpg?_nc_cat=1&_nc_sid=7206a8&_nc_ohc=CD3ImewvDlgAX8f-QvY&_nc_ht=scontent.fsof4-1.fna&oh=782172992e9272da49f35f7f40ef8d63&oe=5EED0201",
        "firstName": "Emil",
        "lastName": "Reinger",
      },
      "status": "accepted",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Friends integration testing should create a notification with status: pending if User A sends a friend request to User B 1`] = `
Object {
  "data": Object {
    "addFriend": Object {
      "action": "Sent you a friend request",
      "creator": Object {
        "avatarImage": "https://scontent.fsof4-1.fna.fbcdn.net/v/t1.30497-1/c71.0.240.240a/p240x240/84241059_189132118950875_4138507100605120512_n.jpg?_nc_cat=1&_nc_sid=7206a8&_nc_ohc=CD3ImewvDlgAX8f-QvY&_nc_ht=scontent.fsof4-1.fna&oh=782172992e9272da49f35f7f40ef8d63&oe=5EED0201",
        "firstName": "Vincent",
        "lastName": "Beahan",
      },
      "notifier": Object {
        "avatarImage": "https://scontent.fsof4-1.fna.fbcdn.net/v/t1.30497-1/c71.0.240.240a/p240x240/84241059_189132118950875_4138507100605120512_n.jpg?_nc_cat=1&_nc_sid=7206a8&_nc_ohc=CD3ImewvDlgAX8f-QvY&_nc_ht=scontent.fsof4-1.fna&oh=782172992e9272da49f35f7f40ef8d63&oe=5EED0201",
        "firstName": "Emil",
        "lastName": "Reinger",
      },
      "status": "pending",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
